executor/README.md

```diff
@@ -0,0 +1,270 @@
+
+#### `PATCH /task/:taskid`
+
+This request also requires a body. Since it is a `PATCH` request, a valid request may contain only some of the following keys:
+
+| key | type   | description                              |
+| -------- | ------ | ---------------------------------------- |
+| `progress` | float  | A number between 0 and 100, representing the progress of the task |
+| `message`   | string | A custom failure reason, expected to be accompanied by a non-zero exit code. |
+| `env`      | object | A map of env variable names to values, a null value will unset the variable. This is really only useful from a before command, since it will only affect commands that run afterward.  |
+
+Example - progress update:
+
+`curl -X PATCH -d '{progress: 10}' $EXECUTOR_ENDPOINT`
+
+Example - setting and unsetting environment variables:
+
+`curl -X PATCH -d '{env: {VAR_A: "123", SOME_B: null}}' $EXECUTOR_ENDPOINT`
+
+Potential responses:
+
+| Code | HTTP Meaning        | Possible reason                          |
+| ---- | ------------------- | ---------------------------------------- |
+| 204  | No Content          | The request was succesfully processed. |
+| 400  | Malformed           | The request body contains an unknown key, or a key's value is the wrong type. |
+| 503  | Service Unavailable | The request was valid, and received, but there was an (ostensibly temporary) failure with mesos itself. |
+
+#### `GET /task/:taskid`
+
+For successful requests, the response body could contain:
+
+| key | type | description |
+|-----|------|-------------|
+| `exit_code` | int | The exit code of the primary command. |
+| `before_codes` | array of ints | The exit codes of the before commands. |
+| `after_codes` | array of ints | The exit codes of the after commands. |
+
+Potential responses:
+
+| Code | HTTP Meaning        | Possible reason                          |
+| ---- | ------------------- | ---------------------------------------- |
+| 200  | Okay          | The task id was valid, and the response body contains task data |
+| 404  | Not found           | The provided task id was unknown or invalid |
+| 503  | Service Unavailable | The request was valid, and received, but there was an (ostensibly temporary) failure with mesos itself. |
+
+### Example use cases
+
+Below are some examples of how to use the above features to solve some common problems.
+
+#### Conditionally preventing command execution
+
+In cases where only one copy of a certain job should be running, before and after commands could be used to acquire and release a lock. Assuming you've written some scripts for managing your lock, you could execute them by passing the additional data to the cook scheduler's job creation API:
+
+```json
+{
+	"command": "my-actual-command",
+	"before_commands": [{
+		"value": "./acquire-zk-lock.sh my-service"
+		"guard": true
+	}],
+	"after_commands":[{
+		"value": "./release-zk-lock.sh my-service"
```

Comment:
Yes, `after_commands` execute even if the main command fails. There's no way to set `guard` for the main command. This was to allow, for example, releasing a lock even if the main command  fails. Additionally, it's possible to have the `after_commands` check the exit code of the main command, in case you want to bail if the main command failed.